### ISSUE TITLE
Upgrade of javascript translations require excessive file operations and are not handled gracefully in GUI

### ISSUE THREAD

**Issue by ianbmacdonald** [woocommerce/woocommerce#28941]
It seems there are a number of  json files (10k+) that seem to be created due to Glotpress translations for Javascript components in woocommerce.  It seems that there is no way to stop these from being generated for all javascript components and pre WC 4.0 and WP 5.6 translations were limited to php components (the .po files).   

So it seems these are a wordpress 5.x feature that does not present until WC > 4.0 that requires translation of all the woocommerce-blocks javascript, and there we have all these .json files being generated.   I found a few links with similar observations 

https://wordpress.org/support/topic/multiple-json-translation-files-2/?view=all
https://wordpress.stackexchange.com/questions/326332/unknown-language-json-files

We like using FTP sockets for plugin upgrades, as it provides a very clear and definite log about plugin upgrade script file operations. Our wordpress has no special language support other than defaults, as we operate primarily in english speaking geography, though our theme includes language support. 

In our ftp logs we see the STOR and CHMOD operations for all of these new translation files. 

Tue Jan 12 17:22:31 2021 [pid 12656] [ftpsecure] FTP response: Client "127.0.0.1", "226 Directory send OK."
Tue Jan 12 17:22:31 2021 [pid 12656] [ftpsecure] FTP command: Client "127.0.0.1", "SITE CHMOD 674 /usr/share/wordpress/wp-content/languages/plugins/woocommerce-zh_TW-wc-admin-app.json"
Tue Jan 12 17:22:31 2021 [pid 12656] [ftpsecure] OK CHMOD: Client "127.0.0.1", "/usr/share/wordpress/wp-content/languages/plugins/woocommerce-zh_TW-wc-admin-app.json 674"
Tue Jan 12 17:22:31 2021 [pid 12656] [ftpsecure] FTP response: Client "127.0.0.1", "200 SITE CHMOD command ok."
Tue Jan 12 17:22:31 2021 [pid 12656] [ftpsecure] FTP command: Client "127.0.0.1", "QUIT"
Tue Jan 12 17:22:31 2021 [pid 12656] [ftpsecure] FTP response: Client "127.0.0.1", "221 Goodbye."

Many of them have very small amounts of information

/var/lib/wordpress/wp-content/languages/plugins# cat woocommerce-zh_TW-fffdec5db595ec0ca7d043c87bae79b8.json
{"translation-revision-date":"2021-01-12 14:46:52+0000","generator":"GlotPress\/3.0.0-alpha.2","domain":"messages","locale_data":{"messages":{"":{"domain":"messages","plural-forms":"nplurals=1; plural=0;","lang":"zh_TW"},"All Reviews":["\u6240\u6709\u8a55\u8ad6"],"List Settings":["\u6e05\u55ae\u8a2d\u5b9a"],"Content":["\u5167\u5bb9"],"Product name":["\u5546\u54c1\u540d\u7a31"]}},"comment":{"reference":"packages\/woocommerce-blocks\/assets\/js\/blocks\/reviews\/all-reviews\/edit.js"}}
 
A recent 4.8 to 4.9 upgrade took 2 hours to deploy all these translation files.  And it seems like a lot of them were touched and updated again when woocommerce related plugins are upgraded. 

It seems that perhaps there should be some consolidation or change in handling of the translations, to reduce the volume of file operations, and the associated time required, for simple plugin upgrades.   DB upgrades have been hived off the plugin upgrade delay in the past -> with a notification that they are occuring in the background as scheduled jobs that can be monitored.   If the approach to translations is not going to change, then this would be a better approach then having the little plugin upgrade icon swirl for two hours, which might other wise cause some users to worry that something has gone wrong if they can not monitor or see the background io operations underway. 

Our instance runs on standard Debian wordpress/apache2/php packages

WooCommerce version:		4.9.0
WooCommerce REST API package:		 4.9.0 /usr/share/wordpress/wp-content/plugins/woocommerce/includes
WooCommerce Blocks package:		 4.0.0 /usr/share/wordpress/wp-content/plugins/woocommerce/packages/woocommerce-blocks/
Action Scheduler package:		 3.1.6 /usr/share/wordpress/wp-content/plugins/woocommerce/packages/action-scheduler
WooCommerce Admin package:		 1.8.3 /usr/share/wordpress/wp-content/plugins/woocommerce/packages/woocommerce-admin
Log directory writable:		 /usr/share/wordpress/wp-content/uploads/wc-logs/
WordPress version:		5.6

Here's an idea of the file counts, 10K new json translations; presumably the file identifiers are stored in a db somewhere, so there is additional table space to consider as well. 

/usr/share/wordpress/wp-content/languages/plugins# ls -al | wc -l
11519

/usr/share/wordpress/wp-content/languages/plugins# ls -al | grep woocommerce | wc -l
10912

/usr/share/wordpress/wp-content/languages/plugins# ls -al | grep woocommerce | grep json | wc -l
10764

/usr/share/wordpress/wp-content/languages/plugins# ls -al | grep woocommerce | grep json | more
-rw-rwxr--+ 1 ftpsecure ftpsecure     553 Jan 12 15:54 woocommerce-ar-00cecdb4dc8514ec7e775622bc723648.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     519 Jan 12 15:54 woocommerce-ar-01739590876f3ab645ec3678e42bbc29.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    2281 Jan 12 15:54 woocommerce-ar-01b768ae1db405a69d13758eda43a58b.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    9519 Jan 12 15:54 woocommerce-ar-02acd74879196411261af7576fdb4b93.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    2162 Jan 12 15:54 woocommerce-ar-0436417c01a5fcf6da938f5082798846.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    1158 Jan 12 15:54 woocommerce-ar-056e3f14efed538fde21950202add8dc.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     806 Jan 12 15:54 woocommerce-ar-0657952e8fe472b7a7f031ef6c2f1e6a.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     622 Jan 12 15:54 woocommerce-ar-09444bf964f9242a7dc345c5da29c237.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     715 Jan 12 15:54 woocommerce-ar-09b8a04693848711e3459fcc6f5f2a94.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    1288 Jan 12 15:54 woocommerce-ar-09c5919a18291dd26d6f52f9d1563a8e.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     712 Jan 12 15:54 woocommerce-ar-0a3b65850c50f16de824a40fa5ff0f0f.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    6927 Jan 12 15:54 woocommerce-ar-0cf7579f6fbc3a394bde81590518d79e.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     822 Jan 12 15:54 woocommerce-ar-0db65848d165e94c1b721836932fa2b2.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    1309 Jan 12 15:54 woocommerce-ar-0dc089127d6d4abad378c797b27b1007.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    9188 Jan 12 15:54 woocommerce-ar-0e3dcd4f3ff5aa786acbbd18de58d420.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     622 Jan 12 15:54 woocommerce-ar-11eaf0a60d602f142e96c14c44445c1c.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    2747 Jan 12 15:54 woocommerce-ar-124cfb7a5d89d9c89c181e1d53e04227.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    1549 Jan 12 15:54 woocommerce-ar-12698f76de1361156158b68e7449b0a9.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    1443 Jan 12 15:54 woocommerce-ar-154eb5541053115f9cf2cc279decf649.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    1093 Jan 12 15:54 woocommerce-ar-15ffb2e3ba27c86287dde1e4045e79cc.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     700 Jan 12 15:54 woocommerce-ar-1605a532b87cc989e30f0c39532c5845.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     658 Jan 12 15:54 woocommerce-ar-16268754aba0e39aeb4f24c835d52820.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     538 Jan 12 15:54 woocommerce-ar-1869430a90f07a7364b4c043829f725f.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    2685 Jan 12 15:54 woocommerce-ar-18a31075fa5f4596edaf76ec78921ff4.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     887 Jan 12 15:54 woocommerce-ar-19f16f4fc69aa97ecde5d39d523dfd8b.json
-rw-rwxr--+ 1 ftpsecure ftpsecure    1114 Jan 12 15:54 woocommerce-ar-19f52ef80d162b0ab094ab96c60a76e8.json
-rw-rwxr--+ 1 ftpsecure ftpsecure     886 Jan 12 15:54 woocommerce-ar-1a349548a3d38de6f8013f63a02ddf4b.json

**Comment by ianbmacdonald** [comment]
Today we tested just upgrading a single plugin,  Google Analytics for WordPress by MonsterInsights.  The plugin was not even active, and the woocommerce javascrip translation updates went into full swing in the background. 

This simple update, of a common, inactive plugin, resulted in an enormous number of file operations and actually still running right now, so will probably complete in an hour or so. 

The larger problem of all of these javascript translation operations, is that there is no safe way to execute an upgrade, with a roll-back that could be completed in a reasonable amount of time to protect pending/oustanding transactions.  If somebody upgrading their site ended up in an unstable state, where they decided to rollback and re-test, they would have to wait for all this woocommerce stuff to finish. 

And this would happen during plugin upgrades that have nothing [that should be] directly linked to the woocommerce website functionality. 

I am not sure how we will handle this risk for now, as once you upgrade into WC >4 and WP >5.6 this seems to just happen, as is our expereince in our various Debian 10 stable  dev/qa/prod environments.   We originally reported this two releases ago with Woocommerce support  in ticket  #3644320 after we determined a 4.9.0 upgrade was generating problems with our upgrade window timing.

Unfortunately they had no input on what or why this might be happening, and suggesting trying to file a bug report if we believed something was not functioning correctly.   We sure hope somebody looks at this soon. 

We are not going to wait the full two hours for the plugin upgrade to complete, but from below you can see all the woo translations being touched by file operations, about 10% (guess)  complete having modified 1800 files so far. 

/usr/share/wordpress/wp-content/languages/plugins# ls -alot | grep woocommerce | grep json | more
-rw-rwxr--+ 1 ftpsecure   66048 Jan 31 14:04 woocommerce-it_IT-wc-admin-app.json
-rw-rwxr--+ 1 ftpsecure   95183 Jan 31 14:03 woocommerce-fa_IR-wc-admin-app.json
-rw-rwxr--+ 1 ftpsecure   56698 Jan 31 14:01 woocommerce-cs_CZ-wc-admin-app.json
-rw-rwxr--+ 1 ftpsecure     488 Jan 31 13:59 woocommerce-zh_TW-fffdec5db595ec0ca
7d043c87bae79b8.json
-rw-rwxr--+ 1 ftpsecure     500 Jan 31 13:59 woocommerce-zh_TW-fe93c2577c317c098
eae3d9773e887a1.json
-rw-rwxr--+ 1 ftpsecure     426 Jan 31 13:59 woocommerce-zh_TW-fdc53b73be01f3738
b23cc383bd00b1a.json
-rw-rwxr--+ 1 ftpsecure     563 Jan 31 13:59 woocommerce-zh_TW-fd20fc8ea14847811
60e8ea0d229d5e9.json
-rw-rwxr--+ 1 ftpsecure     566 Jan 31 13:59 woocommerce-zh_TW-fc4895b6508f32c16
c4d8bc69f8162b6.json
-rw-rwxr--+ 1 ftpsecure     529 Jan 31 13:59 woocommerce-zh_TW-f9c56a0cbb6710505
047e6c835953c82.json
-rw-rwxr--+ 1 ftpsecure     825 Jan 31 13:59 woocommerce-zh_TW-f976da155dfdd0805
88c50ccda5af747.json
-rw-rwxr--+ 1 ftpsecure    1285 Jan 31 13:59 woocommerce-zh_TW-f686a272add740773
cdcb730afdb5454.json
-rw-rwxr--+ 1 ftpsecure     874 Jan 31 13:59 woocommerce-zh_TW-f60049653c419cfb1
e3dc8876e7a8bdb.json
-rw-rwxr--+ 1 ftpsecure    1691 Jan 31 13:59 woocommerce-zh_TW-f56e66fd08c07abc7
8cd844f7e83af8c.json


/usr/share/wordpress/wp-content/languages/plugins# find ./ -mtime -1 -ls | wc -l
1810

**Comment by ianbmacdonald** [comment]
A few more observations;  Upgrading some plugins like "W3 Total Cache" have no impact, whereas other plugins like "Google Analytics for Wordpress MonsterInsights" trigger about 2000 file updates to javascript translations for woocommerce. 

We did the 4.9.2 upgrade in our QA environment, and another 8000 file updates javascript translations were required, taking about and a half. 

After completing the upgrade to 4.9.2 yesterday,  we had a separate upgrade prompt appear today for "Updated Translations" in the translations section of wordpress.  This seemed to trigger yet another update, but only for Woo translations has started the update of the 10k json files again.  A capture below from the GUI shows the process kicking off, 

![Qa_UpdateTranslations](https://user-images.githubusercontent.com/6287157/106500251-96eac000-648f-11eb-90c3-78dc5ea3bb35.png)


It is worth noting that our only installed language is "English (United States)" as shown in the capture below. 

![QA_languages](https://user-images.githubusercontent.com/6287157/106500504-eaf5a480-648f-11eb-86fc-117f43bcfaa1.png)

**Comment by ianbmacdonald** [comment]
With some help from Bill @ Automattic we figured this out.  I'll share some insights he provided. 

It seems that if any point you have .mo files deployed in your language directories, the translation engine will assume you have the language support installed, and just start performing translations. 

https://github.com/WordPress/wordpress-develop/blob/3c40632397cf1ee64e09c7ac767391dfac59e5ea/src/wp-includes/l10n.php#L1360

Recently, the Debian wordpress package added the missing "plugins" and "themes" directories to the languages folder, and in our environments we moved any dangling .mo files into their proper locations after the fix.  We found some plugins had dropped them in the wp-content/plugins and wp-content/themes directories respectively, so we just put them into the proper locations. 

It is worth noting that these directories seem to be missing in a lot of vanilla installs, 

https://github.com/WordPress/wordpress-develop/blob/3c40632397cf1ee64e09c7ac767391dfac59e5ea/src/wp-includes/l10n.php#L1404

Here we see them being added (except core) to the Debian package we use
https://salsa.debian.org/debian/wordpress/-/commit/33338f03f95cade0b78a43e9a17993a0a81c1709

In our case, is is unclear if the language bits we placed into the proper locations were left because of their deployment into incorrect locations previously,  where install scripts did not remove them, or if plugins had deployed them in error, but we did not find them in all our environments.  

For anyone else cleaning up languages that did not have the directories deployed properly earlier, some consideration for installed languages needs to be made as once the files are in their proper locations, the translation engine will 'discover' those languages regardless of if they are installed or not. 

If we were using these languages, we would still hope for some better handling of all the small javascript bits, and still recomment that the translations piece be hived off the plugin installation and/or put into the translations upgrade cleanly and sepearately from the plugin upgrade where each file operation can be seen in the GUI, so a sense of progress is being made. 

Most users will not hang around for unexpectedly long file operations during a plugin upgrade.
